src/App.tsx

```diff
@@ -1,26 +1,80 @@
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 import { Loader, TodoFilter, TodoList, TodoModal } from './components';
+import { useEffect, useMemo, useState } from 'react';
+import { getTodos } from './api';
+import { useAppDispatch, useAppSelector } from './app/hooks';
+import { todosSlice } from './features/todos';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
 
-export const App = () => (
-  <>
-    <div className="section">
-      <div className="container">
-        <div className="box">
-          <h1 className="title">Todos:</h1>
+function getVisibleTodos(todos: Todo[], { query, status }: Filter) {
+  return todos.filter(todo => {
+    const isQuery = todo.title.toLowerCase().includes(query);
+    const isStatus =
+      status === 'all'
+        ? true
+        : status === 'completed'
+          ? todo.completed
+          : !todo.completed;
 
-          <div className="block">
-            <TodoFilter />
-          </div>
+    return isQuery && isStatus;
+  });
+}
+
+export const App = () => {
+  const [todosFromServer, setTodosFromServer] = useState<Todo[]>([]);
+  const [loading, setLoading] = useState(false);
+  const dispatch = useAppDispatch();
+  const { currentTodo, filter } = useAppSelector(state => state);
+
+  const visibleTodos = useMemo(
+    () => getVisibleTodos(todosFromServer, filter),
+    [filter, todosFromServer],
+  );
+
+  useEffect(() => {
+    dispatch(todosSlice.actions.changeTodos(visibleTodos));
+  }, [visibleTodos]);
+
+  useEffect(() => {
+    setLoading(true);
+
+    getTodos()
+      .then(todos => setTodosFromServer(todos))
+      .catch(() => {
+        throw new Error('Todos load fail');
```

Comment:
Throwing an error directly in a promise chain is not recommended as it can lead to unhandled promise rejections. Consider using a more user-friendly error handling approach, such as displaying an error message to the user or logging the error.